managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -2151,7 +2151,13 @@ public void operationFailed(ManagedLedgerException exception) {
 
                 mdEntry.triggerFailed(exception);
             }
-        });
+        };
+
+        if (directPersistMetadataStore) {
+            persistPositionMetaStore(mdEntry, cb);
+        } else {
+            persistPositionToLedger(cursorLedger, mdEntry, cb);
+        }
```

Comment:
Another solution is you can change https://github.com/apache/pulsar/pull/20935/files#diff-fad355f91bd15cc041161f9a46fce62b7fee87fbfb8f0ff8a8b724a1bd1f29eeR3054

to 

```
        if (lh == null) {
            persistPositionMetaStore(mdEntry, callback);
            return;
        }
```

So that you don't need to add param `directPersistMetadataStore` to multiple methods.

Reply:
Good suggestion.

 I have changed the check before `persistPositionToLedger` is executed, which can avoid building the unnecessary `PositionInfo` object. see https://github.com/apache/pulsar/pull/20935/files#diff-fad355f91bd15cc041161f9a46fce62b7fee87fbfb8f0ff8a8b724a1bd1f29eeR2156